### ISSUE TITLE
Using groups in custom loss function

### ISSUE THREAD

**Issue by jrj-d** [dmlc/xgboost#2641]
Hi all,

Is there a plan to make groups available on `DMatrix` so that they can be used in customized objective functions and eval functions?

For instance, using the example in [here](https://github.com/dmlc/xgboost/blob/master/demo/guide-python/custom_objective.py)
```
def logregobj(preds, dtrain):
    labels = dtrain.get_label()
    preds = 1.0 / (1.0 + np.exp(-preds))
    grad = preds - labels
    hess = preds * (1.0-preds)
    return grad, hess
```
could we add a method `dtrain.get_group()`?

Thanks

**Comment by pommedeterresautee** [comment]
Current solution is to use closure in your loss function.
https://github.com/dmlc/xgboost/issues/574#issuecomment-161806666
